src/lib/cardiac_mechanics/LinearSolverUtility.hpp

```diff
@@ -0,0 +1,320 @@
+      // not the initialize method
+      auto derived_ptr =
+          dynamic_cast<dealii::TrilinosWrappers::PreconditionSSOR *>(
+              preconditioner.get());
+      derived_ptr->initialize(matrix);
+      break;
+    }
+    case Preconditioner::SOR: {
+      preconditioner =
+          std::make_unique<dealii::TrilinosWrappers::PreconditionSOR>();
+      // We have to perform a dynamic cast as the PreconditionBase class has
+      // not the initialize method
+      auto derived_ptr =
+          dynamic_cast<dealii::TrilinosWrappers::PreconditionSOR *>(
+              preconditioner.get());
+      derived_ptr->initialize(matrix);
+      break;
+    }
+    case Preconditioner::ILU: {
+      preconditioner =
+          std::make_unique<dealii::TrilinosWrappers::PreconditionILU>();
+      // We have to perform a dynamic cast as the PreconditionBase class has
+      // not the initialize method
+      auto derived_ptr =
+          dynamic_cast<dealii::TrilinosWrappers::PreconditionILU *>(
+              preconditioner.get());
+      derived_ptr->initialize(matrix);
+      break;
+    }
+    case Preconditioner::AMG: {
+      preconditioner =
+          std::make_unique<dealii::TrilinosWrappers::PreconditionAMG>();
+      // We have to perform a dynamic cast as the PreconditionBase class has
+      // not the initialize method
+      auto derived_ptr =
+          dynamic_cast<dealii::TrilinosWrappers::PreconditionAMG *>(
+              preconditioner.get());
+      derived_ptr->initialize(matrix);
+      break;
+    }
+    default: {
+      preconditioner =
+          std::make_unique<dealii::TrilinosWrappers::PreconditionIdentity>();
+      // We have to perform a dynamic cast as the PreconditionBase class has
+      // not the initialize method
+      auto derived_ptr =
+          dynamic_cast<dealii::TrilinosWrappers::PreconditionIdentity *>(
+              preconditioner.get());
+      derived_ptr->initialize(matrix);
+    }
+    }
+  }
+  /**
+   * @brief Solves a linear system using the configured solver and
+   * preconditioner.
+   * @tparam Solver The solver type
+   * @tparam Matrix The lhs matrix type
+   * @tparam Solution The solution vector type
+   * @tparam Vector The rhs vector type
+   * @tparam Preconditioner The preconditioner type
+   * @param solver The solver object.
+   * @param matrix The matrix of the linear system.
+   * @param solution The solution vector.
+   * @param rhs The right-hand side vector.
+   * @param preconditioner The preconditioner object.
+   */
+  template <typename Solver, typename Matrix, typename Solution,
+            typename Vector, typename Preconditioner>
+  void solve(std::unique_ptr<Solver> const &solver, Matrix const &matrix,
+             Solution &solution, Vector const &rhs,
+             std::unique_ptr<Preconditioner> const &preconditioner) {
+    // We have to cast the solver pointer to the derived type as the dealii base
+    // solver class has not the solve method
```

Comment:
Since, unfortunately, the base class does not expose the interface you need, it might make more sense to work directly with the concrete classes, either instantiating the solver on the fly (instead of passing it as argument to the `solve` method - see also my other comment) or making this method a template over the concrete solver type.